cmd/memberagent/main.go

```diff
@@ -184,12 +110,71 @@ func main() {
 	}
 	//+kubebuilder:scaffold:builder
 
-	if err := Start(ctrl.SetupSignalHandler(), &hubConfig, memberConfig, hubOpts, memberOpts); err != nil {
+	if err := Start(ctrl.SetupSignalHandler(), hubConfig, memberConfig, hubOpts, memberOpts); err != nil {
 		klog.ErrorS(err, "problem running controllers")
 		os.Exit(1)
 	}
 }
 
+func buildHubConfig(hubURL string, useCAAuth bool, tlsClientInsecure bool) (*rest.Config, error) {
+	var hubConfig = &rest.Config{
+		Host: hubURL,
+	}
+	if useCAAuth {
+		keyFilePath := os.Getenv("IDENTITY_KEY")
+		certFilePath := os.Getenv("IDENTITY_CERT")
+		if keyFilePath == "" {
+			err := errors.New("identity key file path cannot be empty")
+			klog.ErrorS(err, "error has occurred retrieving IDENTITY_KEY")
+			return nil, err
+		}
+
+		if certFilePath == "" {
+			err := errors.New("identity certificate file path cannot be empty")
+			klog.ErrorS(err, "error has occurred retrieving IDENTITY_CERT")
+			return nil, err
+		}
+		hubConfig.TLSClientConfig.CertFile = certFilePath
+		hubConfig.TLSClientConfig.KeyFile = keyFilePath
+	} else {
+		tokenFilePath := os.Getenv("CONFIG_PATH")
+		if tokenFilePath == "" {
+			err := errors.New("hub token file path cannot be empty if CA auth not used")
+			klog.ErrorS(err, "error has occurred retrieving CONFIG_PATH")
+			return nil, err
+		}
+		err := retry.OnError(retry.DefaultRetry, func(e error) bool {
+			return true
+		}, func() error {
+			// Stat returns file info. It will return
+			// an error if there is no file.
+			_, err := os.Stat(tokenFilePath)
+			return err
+		})
+		if err != nil {
+			klog.ErrorS(err, "cannot retrieve token file from the path %s", tokenFilePath)
+			return nil, err
+		}
+		hubConfig.BearerTokenFile = tokenFilePath
+	}
+
+	hubConfig.TLSClientConfig.Insecure = tlsClientInsecure
+	if !tlsClientInsecure {
+		hubConfig.TLSClientConfig.CAFile = os.Getenv("CA_BUNDLE")
+		hubCA := os.Getenv("HUB_CERTIFICATE_AUTHORITY")
+		if hubCA != "" {
+			caData, err := base64.StdEncoding.DecodeString(hubCA)
+			if err != nil {
+				klog.ErrorS(err, "cannot decode hub cluster certificate authority data")
+				return nil, err
+			}
+			hubConfig.TLSClientConfig.CAData = caData
```

Comment:
This will override the CAFile so it's another place that we want to make sure that users know what they are doing. It's either or.

Reply:
yes, the HUB_CERTIFICATE_AUTHORITY will override the CA_BUNDLE. I do agree that will make confuse for user.

Do you want me to add a validation for that? the validation will check only one of them will be present, otherwise will return a error